packages/ts-plugin/src/frontmatter.ts

```diff
@@ -0,0 +1,150 @@
+	for (const collection of collectionConfigs) {
+		if (collection.config.entries[fsPath]) {
+			return collection.config.entries[fsPath];
+		}
+	}
+}
+
+export function getFrontmatterLanguagePlugin(
+	collectionConfigs: CollectionConfig[],
+): LanguagePlugin<string, FrontmatterHolder> {
+	return {
+		getLanguageId(scriptId) {
+			const fileType = SUPPORTED_FRONTMATTER_EXTENSIONS_KEYS.find((ext) =>
+				scriptId.endsWith(`.${ext}`),
+			);
+
+			if (fileType) {
+				return SUPPORTED_FRONTMATTER_EXTENSIONS[
+					fileType as keyof typeof SUPPORTED_FRONTMATTER_EXTENSIONS
+				];
+			}
+		},
+		createVirtualCode(scriptId, languageId, snapshot) {
+			if (SUPPORTED_FRONTMATTER_EXTENSIONS_VALUES.includes(languageId)) {
+				const fileName = scriptId.replace(/\\/g, '/');
+				return new FrontmatterHolder(
+					fileName,
+					languageId,
+					snapshot,
+					getCollectionName(collectionConfigs, pathToFileURL(fileName).toString().toLowerCase()),
+				);
+			}
+		},
+		updateVirtualCode(_, virtualCode, newSnapshot) {
+			return virtualCode.updateSnapshot(newSnapshot);
+		},
+		typescript: {
+			extraFileExtensions: SUPPORTED_FRONTMATTER_EXTENSIONS_KEYS.map((ext) => ({
+				extension: ext,
+				isMixedContent: true,
+				scriptKind: 7,
+			})),
+			getServiceScript(astroCode) {
+				for (const code of forEachEmbeddedCode(astroCode)) {
+					if (code.id === 'frontmatter-ts') {
+						return {
+							code,
+							extension: '.ts',
+							scriptKind: 3 satisfies ts.ScriptKind.TS,
+						};
+					}
+				}
+				return undefined;
+			},
+		},
+	};
+}
+
+export class FrontmatterHolder implements VirtualCode {
+	id = 'frontmatter-holder';
+	mappings!: CodeMapping[];
+	embeddedCodes!: VirtualCode[];
+	public hasFrontmatter = false;
+
+	lastValidContent: {
+		source: string;
+		generated: string;
+		mappings: CodeMapping[];
+	} = {
+		source: '',
+		generated: '',
+		mappings: [],
+	};
+
+	constructor(
+		public fileName: string,
+		public languageId: string,
+		public snapshot: ts.IScriptSnapshot,
+		public collection: string | undefined,
+	) {
+		this.updateSnapshot(snapshot);
+	}
+
+	updateSnapshot(snapshot: ts.IScriptSnapshot) {
+		this.mappings = [
+			{
+				sourceOffsets: [0],
+				generatedOffsets: [0],
+				lengths: [this.snapshot.getLength()],
+				data: {
+					verification: true,
+					completion: true,
+					semantic: true,
+					navigation: true,
+					structure: true,
+					format: true,
+				},
+			},
+		];
+
+		this.embeddedCodes = [];
+		this.snapshot = snapshot;
+
+		if (!this.collection) {
+			return this;
+		}
+
+		// TODO: More robust frontmatter detection
+		this.hasFrontmatter = this.snapshot.getText(0, 10).startsWith('---');
+
+		const frontmatter = this.hasFrontmatter
+			? this.snapshot
+					.getText(0, this.snapshot.getText(0, this.snapshot.getLength()).indexOf('---', 3) + 3)
+					.replaceAll('---', '   ')
+			: ''; // Generate an empty frontmatter so that we can map an error for a missing frontmatter
```

Comment:
Not sure if it helps, but this is the regex Astro core uses for frontmatter: https://github.com/withastro/astro/blob/a851021498db8ffbb06287c8c3cf1e3fa79407c5/packages/astro/src/vite-plugin-astro/utils.ts#L4

Reply:
Maybe not a lot as I realized this has to work with half-written frontmatter too 🤔

Reply:
No, it'd actually be okay because we can't really know reliably where the frontmatter ends because of YAML's flexibility